crates/runtime/src/materialize/protocol.rs

```diff
@@ -68,35 +84,74 @@ pub fn recv_connector_opened(
 
     let mut checkpoint = db
         .load_checkpoint()
+        .await
         .context("failed to load runtime checkpoint from RocksDB")?;
 
     if let Some(runtime_checkpoint) = runtime_checkpoint {
         checkpoint = runtime_checkpoint.clone();
         tracing::debug!(
-            ?checkpoint,
+            checkpoint=?ops::DebugJson(&checkpoint),
             "using connector-provided OpenedExt.runtime_checkpoint",
         );
     } else {
         *runtime_checkpoint = Some(checkpoint.clone());
         tracing::debug!(
-            ?checkpoint,
+            checkpoint=?ops::DebugJson(&checkpoint),
             "loaded and attached a persisted OpenedExt.runtime_checkpoint",
         );
     }
 
-    Ok((task, accumulator, checkpoint, opened))
+    // Collect all the active journal read suffixes of the checkpoint.
+    let active_read_suffixes: HashSet<&str> = checkpoint
+        .sources
+        .iter()
+        .filter_map(|(journal, _source)| journal.split(';').skip(1).next())
+        .collect();
+
+    // Fetch a persisted max-key value for each active binding.
+    let max_keys: Vec<String> = task.bindings.iter().map(max_key_key).collect();
+    let max_keys = db.multi_get_opt(max_keys, Default::default()).await;
+    let max_keys: Vec<(bytes::Bytes, bytes::Bytes)> = task
+        .bindings
+        .iter()
+        .zip(max_keys.into_iter())
+        .map(|(binding, prev_max)| match prev_max {
+            Ok(None) if active_read_suffixes.get(binding.journal_read_suffix.as_str()).is_some() => {
+                tracing::debug!(state_key=%binding.state_key, "binding has no persisted max-key but is in the runtime checkpoint");
```

Comment:
Possibly in the future, but for the time being we cannot retroactively add this (and don't add it) for materializations which are under way. We _will_ add it for bindings that have their backfill counter bumped. So it would be pretty noisy as a warn or even as an info, because we'll log it for every binding on every restart.